### PR TITLE
Add support to Entry for multiple item hashes

### PR DESCRIPTION
**Pull request by karlbaker02**
This PR adds support for multiple items per entry in the register model. It is a fairly straightforward change, but I will elaborate a little on the modification of the Entry `equals` method.

In order to be able to compare items in the entry, we don't need to worry about what order these item hashes are in. We also allow the possibility of duplicate item hashes per entry, so we can't just store all of these hashes as a set. To get around this problem and not need to re-invent the wheel, I've pulled in the Apache Commons Collection library (http://commons.apache.org/proper/commons-collections/index.html) and am using the method `CollectionUtils.isEqualCollection` to satisfy the above requirements (see the source code at https://commons.apache.org/proper/commons-collections/apidocs/src-html/org/apache/commons/collections4/CollectionUtils.html). If we didn't pull in this library and instead wanted to use the default `List.equals` method that Java provides, we'd have to sort the hashes stored by each entry before doing comparison, as this is how the default method operates. If/when we get to the scale where each entry has many items, this may not be feasible, and is a reason why I went with the Commons Collection library (this by comparison builds up a map of frequencies of each of the items in each of the lists, eliminating the need to sort, but still resulting in an entire pass of each of the lists on each comparison).

Pulling in this library raised a potential security vulnerability during our build process (https://tools.cisco.com/security/center/content/CiscoSecurityAdvisory/cisco-sa-20151209-java-deserialization) but after studying this with Sam, I don't believe this will affect us as we don't allow users to supply serialized data.

Aside from the above, this PR also changes the `equals` and `hashcode` methods of `Entry` to enforce that the fields `key` and `timestamp` should be taken into account when comparing the equality of entries; we were not doing this before. As a consequence, some of the tests have been refactored to accommodate for this.

### CODE REVIEW

File: src/main/java/uk/gov/register/core/Entry.java [hunk 1]
```diff
@@ -16,20 +16,28 @@
...
     private final int entryNumber;
-    private final HashValue hashValue;
+    private final List<HashValue> hashValues;
     private final Instant timestamp;
     private String key;
 
     public Entry(int entryNumber, HashValue hashValue, Instant timestamp, String key) {
         this.entryNumber = entryNumber;
-        this.hashValue = hashValue;
+        this.hashValues = Arrays.asList(hashValue);
```
Comment (johnollier) [inline src/main/java/uk/gov/register/core/Entry.java]: You won't be able to add to the List returned by Arrays.asList(). We should create a new instance of List here.

Comment (karlbaker02) [inline src/main/java/uk/gov/register/core/Entry.java]: Java. I'll change this.

**Comment by michaelabenyohai** [comment]
Can we add some tests for the `equals()` and `hashcode()` methods to `EntryTests`? 

Like - do we consider entries that have the same items but in a different order return `true` for `equals()`? We should work that out and add tests to validate the decision.

**Comment by michaelabenyohai** [comment]
We also need to make sure it's possible for an entry to point to the same item twice (or more). For example, it's possible for more than one entry in the original register to point to the same item, which means that it's possible for an entry in a derivation to point to the same item-hash twice.

Can you add a test that makes sure it's possible for an entry to contain the same item-hash multiple times?

**Pull request merged**